twilight-gateway/src/shard.rs

```diff
@@ -0,0 +1,993 @@
+            heartbeat_interval: None,
+            id: shard_id,
+            latency: Latency::new(),
+            ratelimiter: None,
+            session,
+            status: ConnectionStatus::Connected,
+            user_channel: MessageChannel::new(),
+        })
+    }
+
+    /// Immutable reference to the configuration used to instantiate this shard.
+    pub const fn config(&self) -> &Config {
+        &self.config
+    }
+
+    /// ID of the shard.
+    pub const fn id(&self) -> ShardId {
+        self.id
+    }
+
+    /// Whether the shard is currently connected to the gateway.
+    ///
+    /// The shard may not be connected if the gateway session was recently
+    /// invalidated and has not yet reconnected, or if the shard was explicitly
+    /// [closed] by the user.
+    ///
+    /// [closed]: Self::close
+    pub const fn status(&self) -> &ConnectionStatus {
+        &self.status
+    }
+
+    /// Shard latency statistics, including average latency and recent heartbeat
+    /// latency times.
+    pub const fn latency(&self) -> &Latency {
+        &self.latency
+    }
+
+    /// Statistics about the number of available commands and when the command
+    /// ratelimiter will refresh.
+    ///
+    /// This won't be present if ratelimiting was disabled via
+    /// [`ConfigBuilder::ratelimit_messages`].
+    ///
+    /// [`ConfigBuilder::ratelimit_messages`]: crate::ConfigBuilder::ratelimit_messages
+    pub const fn ratelimiter(&self) -> Option<&CommandRatelimiter> {
+        self.ratelimiter.as_ref()
+    }
+
+    /// Immutable reference to the active gateway session.
+    ///
+    /// An active session may not be present if the shard has recently
+    /// disconnected or had its session invalidated and has not yet performed a
+    /// reconnect.
+    pub const fn session(&self) -> Option<&Session> {
+        self.session.as_ref()
+    }
+
+    /// Wait for the next Discord event from the gateway.
+    ///
+    /// # Errors
+    ///
+    /// Returns a [`ReceiveMessageErrorType::Deserializing`] error type if the
+    /// message payload failed to deserialize.
+    ///
+    /// Returns a [`ReceiveMessageErrorType::FatallyClosed`] error type if the
+    /// shard was closed due to a fatal error, such as invalid authorization.
+    ///
+    /// Returns a [`ReceiveMessageErrorType::Process`] error type if the shard
+    /// failed to internally process a received event.
+    ///
+    /// Returns a [`ReceiveMessageErrorType::Reconnect`] error type if the shard
+    /// failed to reconnect to the gateway. This isn't a fatal error and can be
+    /// retried.
+    ///
+    /// Returns a [`ReceiveMessageErrorType::SendingMessage`] error type if the
+    /// shard failed to send a message to the gateway, such as a heartbeat.
+    pub async fn next_event(&mut self) -> Result<Event, ReceiveMessageError> {
+        let mut bytes = loop {
+            match self.next_message().await? {
+                Message::Binary(bytes) => break bytes,
```

Comment:
Is it not faulty of us to try to deserialize binary messages? The sender would have it as `Text` if it was valid UTF-8 (JSON must be UTF-8).

Reply:
Discord sends binary websocket messages